rootston/desktop.c

```diff
@@ -609,6 +613,51 @@ static void handle_layout_change(struct wl_listener *listener, void *data) {
 	}
 }
 
+#ifdef WLR_HAS_XWAYLAND
+
+static void handle_xwayland_keyboard_grab_destroy(struct wl_listener *listener,
+		void *data) {
+	struct roots_xwayland_keyboard_grab_v1 *roots_grab =
+		wl_container_of(listener, roots_grab, destroy_listener);
+
+	wl_list_remove(&roots_grab->link);
+	free(roots_grab);
+}
+
+static void handle_xwayland_keyboard_grab(struct wl_listener *listener, void *data) {
+	struct roots_desktop *desktop =
+		wl_container_of(listener, desktop, xwayland_keyboard_grab_v1_listener);
+	struct wlr_xwayland_keyboard_grab_v1_grab *grab = data;
+
+	struct roots_view *found = NULL;
+	struct roots_view *view;
+	wl_list_for_each(view, &desktop->views, link) {
+		if (view->wlr_surface == grab->surface) {
+			found = view;
+			break;
+		}
+	}
+
+	if (found) {
+		struct roots_seat *seat = grab->seat->data;
+		roots_seat_set_focus(seat, found);
+
+		struct roots_xwayland_keyboard_grab_v1 *roots_grab =
+			calloc(1, sizeof(struct roots_xwayland_keyboard_grab_v1));
+		if (!roots_grab) {
+			return;
+		}
+
+		roots_grab->grab = grab;
+		wl_signal_add(&grab->events.destroy, &roots_grab->destroy_listener);
+		roots_grab->destroy_listener.notify = handle_xwayland_keyboard_grab_destroy;
+
+		wl_list_insert(&seat->xwayland_keyboard_grabs, &roots_grab->link);
```

Comment:
We probably don't want compositors to change the list manually. I'm also not sure if adding fields like `xwayland_keyboard_grabs` to the seat for extensions is fine.